### ISSUE TITLE
Bignum support

### ISSUE THREAD

**Issue by akheron** [akheron/jansson#69]
Add bigint/bigfloat types and hooks to allow using external libraries to manage them.

There's an initial patch by Deron Meranda here: http://deron.meranda.us/computing/c/jansson-bignum/

**Comment by sakubara** [comment]
I would like to bring this topic up again since I see need for it and also have a solution in mind. I also would like to provide the patch for it, if desired. This would also help to solve #288.

I would like to go a different route though compared to the way it was suggested in the patch linked above or the PR #158. I would recommend to store integers and floating point values internally as string values just as how they were extracted from the Json input. This would retain the original value and help in round trip decoding and encoding. The machine representation of the value can be stored alongside the string representation, if performance is a concern.

Users of the jansson can decide to use the string representation of these values and feed them to libraries like GMP in order to process them.

The API provided by jansson can be kept in its current state, the conversion from string to integer and back again would happen on demand and hidden from the user of the API. A new function `int json_integer_value(const json_t *integer, json_int_t *out)` could be introduced that returns `-1` whenever the stored integer overflows a machine word/`json_int_t`. The function `strtoll` provides this information, so this can be reused and implemented easily.

The biggest upside would be that this eliminates the need for the extra types `JSON_DECIMAL` or `JSON_BIGINTEGER` and `JSON_BIGREAL`. This would relieve users of jansson from the burden to test and handle for different types of integer values which essentially mean the same thing.

**Comment by akheron** [comment]
Sounds good. A few notes:

 You would also need to add a function to extract the original string representation for integers and reals.

 Storing the string representation should be optional, so that it's only stored if a certain decoding flag is given. This would lower the memory impact for those who don't use the original string.

`json_real_t` and `json_integer_t` structs could be expanded by adding a `char str[]` field to the end, and only store a single `\0` if the user didn't opt in for storing the original string. This way, the memory impact for those who don't need the string representation would be minimal.

Looking at the code, `json_string_t` could also use this method to avoid an extra `malloc()` call for the string data.

**Comment by sakubara** [comment]
> You would also need to add a function to extract the original string representation for integers and reals.

Of course, the changes would be pointless without access to these values. I'm aware that the internals of the `json_t` type are handled as an opaque data structure and access to the information held by the objects is granted by the respective functions which form the interface to this library.

> Storing the string representation should be optional, so that it's only stored if a certain decoding flag is given. This would lower the memory impact for those who don't use the original string.

This is a very good idea, I'll implement it that way. The functions to retrieve the string representation of the stored numbers would do a string conversion on demand in this case then. Is this a good idea?

> json_real_t and json_integer_t structs could be expanded by adding a `char str[]` field to the end, […]

I'm afraid this is not possible since `json_t` objects are mutable by design. Changing the value of a `json_t` object with the `json_*_set` family of functions may need to reallocate the data structures, if the new value exceeds the current objectk in sice, which in turn requires the API to return a pointer to the (new) `json_t` object.

**Comment by akheron** [comment]
>I'm afraid this is not possible since `json_t` objects are mutable by design. Changing the value of a `json_t` object with the `json_*_set` family of functions may need to reallocate the data structures, if the new value exceeds the current object in size, which in turn would require the API to return a pointer to the (new) `json_t` object.

Ah, true.

**Comment by cathay4t** [comment]
How about we use the yajl way:

`yajl_gen_number (yajl_gen hand, const char *num, unsigned int len)`

which allow user generate `json_t` using string of number:

`json_t *json_number(const char *num_str)`.

**Comment by Ghazibelaam** [comment]
Anything has been done yet to support Unsigned 64 bit numbers?

**Comment by lano1106** [comment]
I got this error while parsing a JSON string: too big integer near '15726282638479065023'

This number is bigger than LLONG_MAX but is smaller than ULLONG_MAX so it would fit into an unsigned long long...

I could definitely benefit from ull support...
